traversal/scanner/GraphIterator.java

```diff
@@ -377,18 +377,70 @@ private void reset() {
+            edgeInputs.forEach((edge, vertex) -> {
+                if (edge.isRolePlayer()) {
+                    overlappingRPEdges.computeIfAbsent(vertex, v -> new ArrayList<>()).add(edge.asRolePlayer());
+                } else otherEdges.add(edge);
+            });
+            return intersectEdges(overlappingRPEdges, otherEdges);
+        }
+
+        private Forwardable<Vertex<?, ?>, Order.Asc> intersectEdges(Map<Vertex<?, ?>, List<ProcedureEdge.Native.Thing.RolePlayer>> rpEdges,
+                                                                    List<ProcedureEdge<?, ?>> otherEdges) {
+            List<Forwardable<Vertex<?, ?>, Order.Asc>> iterators = new ArrayList<>();
+            otherEdges.forEach(e -> iterators.add(branch(edgeInputs.get(e), e)));
+            rpEdges.forEach((vertex, edges) -> {
+                if (edges.size() == 1) {
+                    iterators.add(branch(edgeInputs.get(edges.get(0)), edges.get(0))
+                            .mapSorted(KeyValue::key, thing -> KeyValue.of(thing.asThing(), null), ASC));
+                } else iterators.add(intersectOverlapping(edges));
+            });
+            if (iterators.size() == 1) return iterators.get(0);
+            else return intersect(iterate(iterators), ASC);
         }
 
-        private Optional<Vertex<?, ?>> lastIntersection() {
+        /**
+         * Overlapping RolePlayer edges (ie. from the same vertex are either):
+         * identical -- (upper: $x, upper: $y); $x is $y;
+         * strict subsets/supersets of each other -- (role: $x, upper: $y); $x is $y; -- solved by sorting most restrictive iterators first
+         * completely disjoint -- (upper: $x, lower: $x); $x isa $y;
+         * It is impossible for two RolePlayer edges to return partially overlapping iterators.
+         */
+        private Forwardable<Vertex<?, ?>, Order.Asc> intersectOverlapping(List<ProcedureEdge.Native.Thing.RolePlayer> edges) {
+            assert edges.size() > 1 && iterate(edges).map(edgeInputs::get).toSet().size() == 1;
+            ArrayList<Forwardable<Vertex<?, ?>, Order.Asc>> iterators = new ArrayList<>(edges.size());
+            edges.stream().sorted(Comparator.comparing(e -> e.roleTypes().size()))
+                    .forEach(e -> {
+                        // rare cases have RP edges in different directions from the same vertex, so look up scope each time
+                        Scopes.Scoped scoped = scopes.getOrInitialise(e.scope());
+                        iterators.add(branch(edgeInputs.get(e), e)
+                                .filter(thingAndRole -> isRoleAvailable(thingAndRole.value(), e, scoped, edges))
+                                .mapSorted(KeyValue::key, v -> KeyValue.of(v.asThing(), null), ASC));
+                    });
+            // WARN: we rely on the intersection operating in order of the iterators provided...
+            return intersect(iterate(iterators), ASC);
+        }
```

Comment:
the current solution to handling role player edges from the same start.
We have definitely handled the `identical` and `disjoint` cases perfectly with this approach. However, the subset/superset is only iffy since we rely on the intersection operating a specific way...

we can improve the solution for this case somehow...

Reply:
To summarise the in-person discussion: the subset case relies on the intersection iterator advancing the iterators in the order provided within the intersection. Put simply, we want to exclude the subset's values from the superset, not the other way around. This is the cleanest solution we came up with in the course of the discussion.